[code-completion] Suggest module names, and handle module name bases for member completion in type position

nathawes opened the pull request:
Previous we didn't suggest modules for top-level completions, and for module type member completions we would just return if the parsed type loc didn't type-check as a type.

Resolves rdar://problem/56800640
Resolved rdar://problem/56801340

Commit:
[code-completion] Complete after qualifying module names in type position

Previously we would just return if the parsed type loc didn't type check as a
type.

nathawes commented:
@swift-ci please test

rintaro commented on test/IDE/complete_from_swift_module.swift:
```diff
@@ -99,3 +105,20 @@ func testPostfixOperator1(x: Int) {
+// STDLIB_TYPE_QUALIFIED: Decl[TypeAlias]/CurrNominal: Index[#Int#]; name=Index
+// STDLIB_TYPE_QUALIFIED: Decl[TypeAlias]/CurrNominal: Element[#Element#]; name=Element
+// STDLIB_TYPE_QUALIFIED: Keyword/None: Type[#Array.Type#]; name=Type
+// STDLIB_TYPE_QUALIFIED: End completions
```
I believe this test case works even before this change.
Could you add a test for `Swift.#^TOKEN^#` at type position as well?

nathawes replied:
Yeah, it was already working. Added the extra test.

Commit:
[code-completion] Suggest module names in type and expression position

They can be used to qualify both.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

rintaro commented on test/SourceKit/CodeComplete/complete_sort_order.swift:
```diff
@@ -33,6 +33,8 @@ func test() {
+// CONTEXT-NOT: key.name:
 // CONTEXT: key.name: "foo(a:)"
 // CONTEXT-NOT: key.name:
 // CONTEXT: key.name: "foo(a:)"
```
`complete_sort_order` is the module name of the current file. It's a little strange if it precedes other decls in the module.
It probably makes sense if `Decl[Module]` has lower priority than other decls.
@benlangmuir what do you think?

benlangmuir replied:
I agree

Commit:
[code-completion] Give Modules a SemanticContextKind of None

This seems more correct than OtherModule, and means they're treated as
low-priority results for the session-based completion APIs. With default
options that means won't be shown in global completion results but will be
once there's some matching filter text. It also sorts them below any symbols
that actually come from other modules, which is desirable, since most symbols
don't need to be module qualified.

The SemanticContextKind change doesn't seem to affect Xcode's handling of the
results.

nathawes commented:
@swift-ci please test

rintaro reviewed (approved):
LGTM!
